crates/subspace-farmer/src/ws_rpc_server.rs

```diff
@@ -430,17 +426,23 @@ impl RpcServerImpl {
     /// Read and decode the whole piece
     async fn read_and_decode_piece(&self, piece_index: PieceIndex) -> Result<Piece, Error> {
         let piece_fut = tokio::task::spawn_blocking({
-            let plot = self.plot.clone();
+            let plots = self.plots.clone();
 
-            move || plot.read(piece_index)
+            move || {
+                plots.iter().find_map(|plot| {
+                    plot.read(piece_index)
+                        .ok()
+                        .map(|piece| (piece, plot.address()))
+                })
+            }
         });
-        let mut piece = piece_fut.await.unwrap().map_err(|error| {
-            debug!("Failed to read piece with index {}: {}", piece_index, error);
+        let (mut piece, address) = piece_fut.await.unwrap().ok_or_else(|| {
+            debug!("Failed to read piece with index {}", piece_index);
 
             Error::Custom("Object mapping found, but reading piece failed".to_string())
         })?;
 
-        self.subspace_codec
+        SubspaceCodec::new(&address)
```

Comment:
Same here, we don't want to instantiate codec on every call, needs a TODO here and below.